prover/crates/bin/prover_autoscaler/src/global/scaler.rs

```diff
@@ -0,0 +1,344 @@
+use std::{collections::HashMap, str::FromStr};
+
+use chrono::Utc;
+use debug_map_sorted::SortedOutputExt;
+use regex::Regex;
+
+use super::{queuer, watcher};
+use crate::{
+    cluster_types::{Cluster, Clusters, PodStatus, GPU},
+    metrics::AUTOSCALER_METRICS,
+    task_wiring::Task,
+};
+
+const DEFAULT_SPEED: u64 = 500;
+
+#[derive(Default, Debug, PartialEq, Eq)]
+struct GPUPool {
+    name: String,
+    gpu: GPU,
+    provers: HashMap<PodStatus, u32>, // TODO: consider using i64 everywhere to avoid type casts.
+    preemtions: u64,
+    max_pool_size: u32,
+}
+
+impl GPUPool {
+    fn sum_by_pod_status(&self, ps: PodStatus) -> u32 {
+        self.provers.get(&ps).cloned().unwrap_or(0)
+    }
+
+    fn to_key(&self) -> GPUPoolKey {
+        GPUPoolKey {
+            cluster: self.name.clone(),
+            gpu: self.gpu,
+        }
+    }
+}
+
+#[derive(Debug, Eq, Hash, PartialEq)]
+struct GPUPoolKey {
+    cluster: String,
+    gpu: GPU,
+}
+
+pub struct Scaler {
+    /// namespace to Protocol Version configuration.
+    namespaces: HashMap<String, String>,
+    watcher: watcher::Watcher,
+    queuer: queuer::Queuer,
+
+    /// Which cluster to use first.
+    cluster_priorities: HashMap<String, u32>,
+    prover_speed: HashMap<GPU, u64>,
+    // `gpu` group must be specified.
+    prover_deployment_re: Regex,
+    prover_pod_re: Regex,
+}
+
+impl Scaler {
+    pub fn new(watcher: watcher::Watcher, queuer: queuer::Queuer) -> Self {
+        Self {
+            namespaces: HashMap::from([
+                ("prover-red".to_string(), "0.24.2".to_string()),
+                ("prover-blue".to_string(), "0.24.1".to_string()),
+            ]), // TODO: read from config.
+            watcher,
+            queuer,
+
+            // TODO: move to config.
+            cluster_priorities: HashMap::from([
+                ("mainnet2".to_string(), 0),
+                ("mainnet1-use4".to_string(), 10),
+            ]),
+            // TODO: add config.
+            prover_speed: HashMap::from([(GPU::L4, 500)]),
+            prover_deployment_re: Regex::new(
+                r"^prover-gpu-fri-spec-(\d{1,2})?(-(?<gpu>[ltvpa]\d+))?$",
+            )
+            .unwrap(),
+            prover_pod_re: Regex::new(r"^prover-gpu-fri-spec-(\d{1,2})?(-(?<gpu>[ltvpa]\d+))?")
+                .unwrap(),
+        }
+    }
+
+    fn convert2gpu_pool(&self, namespace: &String, cluster: &Cluster) -> Vec<GPUPool> {
```

Comment:
Let's use full name of the method: `convert_to_gpu_pool`

Reply:
Renamed